rfc/distributed-tracing.md

```diff
@@ -0,0 +1,138 @@
+
+#### Non-Goals
+
+- Bundling a trace aggregator with Spinnaker.
+
+## Motivation and Rationale
+
+Spinnaker is a complicated system to operate. Consisting of approximately 10 services,
+it can be extremely difficult to follow the lifecycle of a request without intimate knowledge
+of the interdependencies between each service. This makes troubleshooting issues painful,
+as it often requires jumping between multiple log streams to follow a request as it moves
+through the system.
+
+Distributed tracing is a method for following the lifecycle of a request as it moves through
+various connected systems. Each unique request is "tagged" with an ID where it originates, and this
+ID is then propagated alongside any downstream requests.
+
+When combined with a log aggregator of your choice, including these trace IDs in logs enables
+querying for a single ID to identify all logs across all services involved in the request.
+
+Furthermore, traces can be pushed to an aggregator such as Zipkin to enable visualisation
+of the request lifecycle.
+
+## Timeline
+
+I propose that this work be undertaken in three distinct stages:
+
+### Stage One
+
+Introduce a standard `logback-config.xml` in Kork for all services to consume. Currently, each service
+defines it's on Logback configuration in the `{service}-web` module. These are all nearly identical, with
+minute, inconsequential differences. By introducing a standard logging configuration it becomes easier to
+make broad improvements to logging.
+
+This will be achieved by removing the `logback-defaults.xml` file from each service and including the Kork
+equivalent in each service's source set, as described [here](https://stackoverflow.com/a/43686346).
+
+### Stage Two
+
+Introduce a set of libraries as dependencies in Kork, and wire them up to OkHttp and logging. This RFC proposes
+using [Micrometer Tracing](https://micrometer.io/docs/tracing), the [successor](https://spring.io/projects/spring-cloud-sleuth)
+to Spring Sleuth.
+
+The Micrometer Tracing dependencies will be added to Kork as below:
+
+```text
+implementation platform('io.micrometer:micrometer-tracing-bom:${version}')
+implementation 'io.micrometer:micrometer-tracing:${version}'
+```
+
+The next step will be to choose between [OpenTelemetry](https://opentelemetry.io/docs/concepts/signals/traces/) and
+[Brave](https://zipkin.io/pages/tracers_instrumentation.html). Brave was added to Spinnaker in
+[kork#744](https://github.com/spinnaker/kork/pull/774) and is already wired up to OkHttp.
+
+Both [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/okhttp/okhttp-3.0/library)
+and [Zipkin](https://github.com/openzipkin/brave/blob/master/instrumentation/okhttp3) provide instrumentation for OkHttp,
+therefore we are free to choose. OpenTelemetry has been gaining a lot of momentum, but Brave is also widely used.
+I leave this choice for discussion in this RFC.
```

Comment:
I'm beginning to get educated here and am leaning to OpenTelemetry.  Need to get more educated to form a proper preference.